### PR TITLE
#181414762 (feature): Send a reset password email 

### PR DESCRIPTION
**Pull request by niyojuste**
### What does this PR do?
 Resets the password of a user
### Description of Task to be completed?
- Send a reset password email with a token expiring after 5 minutes
- Update a user's password

### How should this be manually tested?
Get started By Running commands bellow:
  - clone repo
  - switch to ft-reset-password-181414762 branch
  - ``yarn install``: to install all dependencies
  - Create .env file
  - add environment variables as per .env.example

Run the test suite:
  - ``yarn test``: to run the test suite

Run the application locally:
- ``yarn dev``: to start the application in development mode
- open the application in your browser(you will be redirected to the documentation) and try these requests:
  - **/api/users/send/forgot-password/{email}**
  - **/api/users/reset-password/{id}/{token}** or **open the email used and follow the link to reset the password**
### Any background context you want to provide?
- Service API for sending emails of choice is Google
### What are the relevant pivotal tracker stories?
[#181414762](https://www.pivotaltracker.com/story/show/181414762)
### Questions:

### CODE REVIEW

File: src/controllers/auth.js [hunk 1]
```diff
@@ -13,14 +13,8 @@ const { signToken, checkPassword } = encryption;
...
     try {
-      const { email, password } = req.body;
+      const foundUser = req.user;
 
-      const foundUser = await findUser({ email });
-
-      const isMatch = checkPassword(password, foundUser.password);
-      if (!isMatch) {
-        return ErrorResponse.notFoundError(res, 'Invalid credentials');
-      }
```
Comment (niyojuste) [inline src/controllers/auth.js:23]: This was already being handled in the verifyLogin middleware in middlewares/login.js

File: src/controllers/user.js [hunk 2]
```diff
@@ -57,65 +56,6 @@ class UserController {
...
-
-      if (!email.envelope) {
-        return res.status(502).json({ error: 'Bad Gateway' });
-      }
-      res.json(email);
-    } catch (error) {
-      res.status(404).json({ error: error.message });
-    }
-  };
-
```
Comment (niyojuste) [inline src/controllers/user.js:118]: Moved sending email handlers in EmailController.js to reduce lines of code in this file and separate the logic from handlers that interact with the user directly

File: views/reset-password.ejs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+<body>
+  <form method="POST">
+    <label for="password">New Password</label>
+    <input type="password" name="password" placeholder="Enter a new password">
+    <label for="confirmPassword">Confirm Password</label>
+    <input type="password" name="confirmPassword" placeholder="Repeat the new password">
+    <button type="submit">Reset Password</button>
+  </form>
+</body>
+</html>
```
Comment (ericus123) [inline views/reset-password.ejs:18]: Password reset functionalities can be easily tested on the backend.
This page will be designed on the Frontend.
Remove it

**Review by ericus123 — CHANGES_REQUESTED** [review]

File: views/reset-password.ejs (hunk 3 above)
Comment (ericus123) [inline views/reset-password.ejs]: @niyojuste is it for just demonstration ? if so you can leave it but otherwise it can be unnecessary

**Comment by ericus123** [comment]
@niyojuste I'm able to rest my password to be the same as the old password.
Change that. The new password set should never be the same as old one

**Review by niyojuste — COMMENTED** [review]
@ericus123 I changed the response to return a message informing the link is valid and proceeded to remove the input page 

I also took care of the new password matching the old password issue

File: src/validations/newPasswordValidation.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      return ErrorResponse.semanticError(
+        res,
+        `${error.message.replace(/["\\]/g, '')}`
+      );
+    }
+
+    if (Protection.checkPassword(req.body.password, req.user.password)) {
+      return ErrorResponse.semanticError(
+        res,
+        'New password can not be the same as current password'
```
Comment (ericus123) [inline src/validations/newPasswordValidation.js:40]: say **old password** instead of **current password**

**Review by ericus123 — APPROVED** [review]

File: src/validations/newPasswordValidation.js (hunk 4 above)
Comment (itisWasp) [reply src/validations/newPasswordValidation.js:40]: @niyojuste Improve the error message as ``Please try a different password you have not used``.

**Review by Robertishimwe — APPROVED** [review]

**Review by didasmbalanya — APPROVED** [review]

**Pull request merged**